lib/routemaster/fetcher.rb

```diff
@@ -28,23 +27,45 @@ def initialize(middlewares: [], listener: nil)
     # string otherwise.
     def get(url, params: {}, headers: {})
       host = URI.parse(url).host
-      connection.get(url, params, headers.merge(auth_header(host)))
+      response_wrapper do
+        connection.get(url, params, headers.merge(auth_header(host)))
+      end
+    end
+
+    def post(url, body: {}, headers: {})
+      host = URI.parse(url).host
+      response_wrapper do
+        connection.post do |req|
+          req.url url
+          req.headers = headers.merge(auth_header(host))
+          req.body = body.to_json
+        end
+      end
+    end
+
+    def discover(url)
+      get(url)
     end
 
     private
 
-    def generate_middlewares(middlewares)
-      default_middleware = [DEFAULT_MIDDLEWARE, listener: @listener]
-      middlewares.unshift(default_middleware)
+    def response_wrapper(&block)
+      response = block.call
+      @response_class ? @response_class.new(response) : response
```

Comment:
I'm guessing you need to inject `self` here:

```ruby
@response_class ? @response_class.new(response, client: self) : response
```

so that the materialised resource knows which client to call when walking links.